core/src/main/java/feign/ClientInterceptor.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2012-2022 The Feign Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package feign;
+
+/**
+ * Zero or One {@code ClientInterceptor} may be configured for purposes such as tracing - mutate
+ * headers, execute the call, parse the response and close any previously created objects.
+ */
+public interface ClientInterceptor {
+
+  ClientInterceptor DEFAULT = new ClientInterceptor() {
+    @Override
+    public void beforeExecute(ClientInvocationContext context) {}
```

Comment:
I really don't like `before`/`after` interceptors, exactly cause they can't share information and then the API need a mutable context to address it.

If you look at other `Interceptors` feign has, it's an `around` interceptor.  That gives the implementing code more flexibility to share data w/o the burden of having a mutable holder.

Another weak argument I have to use an around style is that java did that for servlet filters, so, can't be that bad =)

If you really want to expose a before/after style API, I guess we can have `ClientInterceptor` with the `around` method and then a `ClientInterceptorAdapter` with before/after methods.

Reply:
Just to be clear, I have no problems with the context parameter, my problem is with the mutable holder.

I know feign has a lot of mutable states, and if I could I would eliminate them all, but, legacy can be hard.

Reply:
Absolutely that's not a problem to do it with an `around` mechanism. The only concern I have is that with `before` and `after` you can have a list of these. If you have `around` you can't really have a list of those (unless I'm missing sth)

Reply:
Actually I did manage to rewrite it to use around. Check the latest commit.